src/sync.jl

```diff
@@ -0,0 +1,131 @@
+module Sync
+
+    @enum SyncMode mutex pin disabled
+
+    const gap_lock = ReentrantLock()
+    const pinned_thread = Ref{Int}(1)
+
+    @inline function lock()
+        Base.lock(gap_lock)
+    end
+
+    @inline function unlock()
+        Base.unlock(gap_lock)
+    end
+
+    @inline function check_lock()
+        @assert gap_lock.locked_by === Base.current_task()
+    end
+
+    @inline function check_pinned()
+        @assert Threads.threadid() == pinned_thread[]
+    end
+
+    # To switch between multi-threaded and single-threaded mode, we
+    # define functions that install the appropriate handlers for sync()
+    # etc.
+    #
+    # This is necessary because otherwise precompilation would fix
+    # the mode at whatever state it was during precompilation. Thus,
+    # initially loading GAP.jl in single-threaded mode would also keep
+    # synchronization off in multi-threaded mode.
+    #
+    # However, Julia tracks function dependencies. If a function changes
+    # upon which another depends, both are being recompiled. Thus,
+    # by installing the proper version during __init__(), we force
+    # selective recompilation of the affected functions as needed.
+
+    function _mode_mutex()
+        Sync.eval(:(@inline function sync(f::Function)
+            try
+                lock()
+                f()
+            finally
+                unlock()
+            end
+        end))
+        Sync.eval(:(@inline function sync_noexcept(f::Function)
+            lock()
+            t = f()
+            unlock()
+            t
+        end))
+        Sync.eval(:(@inline function check_sync(f::Function)
+            check_lock()
+            f()
+        end))
+    end
+
+    function _mode_pinned()
+        Sync.eval(:(@inline function sync(f::Function)
+            check_pinned()
+            f()
+        end))
+        Sync.eval(:(@inline function sync_noexcept(f::Function)
+            check_pinned()
+            f()
+        end))
+        Sync.eval(:(@inline function check_sync(f::Function)
+            check_pinned()
+            f()
+        end))
+    end
+
+    function _mode_nosync()
+        Sync.eval(:(@inline function sync(f::Function)
+            f()
+        end))
+        Sync.eval(:(@inline function sync_noexcept(f::Function)
+            f()
+        end))
+        Sync.eval(:(@inline function check_sync(f::Function)
+            f()
+        end))
+    end
+
+    function switch(mode::SyncMode)
```

Comment:
Does this function exist only for debugging / benchmarking? Or whatever else is its purpose? In either case, a brief source comment indicating its purpose would be nice.

Reply:
The goal here is to let an (experienced) user switch between locking modes. By default, this pins the GAP interpreter to the main thread. However, one can then also switch to full mutex mode in case GAP needs to be accessed from more than one thread, or to turn off locking in case somebody wants to avoid the overhead involved and takes care of synchronization in a different fashion, e.g. via using `@sync` directly.

I agree that there should be a comment and will be adding one. I'll probably also rename `@sync`, since Julia's `Base.@sync` has different semantics and that would be confusing.